python/wakapi_anyide/runner.py

```diff
@@ -35,12 +35,15 @@ async def heartbeat_task(env: Environment, queue: Queue[Event], watchers: Sequen
     while not should_shutdown.is_set():
         changed_events: Dict[str, Event] = dict()
         
+        # wait until the next heartbeat due, processing events in the meantime
         while (due := next_heartbeat_due - time.time()) > 0 and not should_shutdown.is_set():
+            # we don't want to block on getting an event if we're already processing one, so only get a new event if the previous one is done
+            # I'm not sure exactly what this does to prevent thread blocking
```

Comment:
It does not. The purpose is to wait for the future if it has not already completed, as otherwise we'd be dropping events